src/rules/wf/test.rs

```diff
@@ -519,3 +519,45 @@ fn higher_ranked_cyclic_requirements() {
         }
     }
 }
+
+#[test]
+fn higher_ranked_inline_bound_on_gat() {
+    lowering_success! {
+        program {
+            trait Fn<T> { }
+            struct Ref<'a, T> { }
+            struct i32 {}
+
+            struct fn<T> { }
+
+            impl<'a, T> Fn<Ref<'a, T>> for for<'b> fn<Ref<'b, T>> { }
+
+            trait Bar {
+                type Item<T>: forall<'a> Fn<Ref<'a, T>>;
+            }
+
+            impl Bar for i32 {
+                type Item<T> = for<'a> fn<Ref<'a, T>>;
+            }
+        }
+    }
+
+    lowering_error! {
+        program {
+            trait Fn<T, U> { }
+            struct i32 {}
+
+            struct fn<T, U> { }
+
+            trait Bar {
+                type Item<T>: forall<U> Fn<T, U>;
```

Comment:
Can we get a test where the bound *would've* worked, but for the higher-ranked nature of it? I am thinking of something like:

```
struct foo;
impl Fn<i32, i32> for foo { }

impl Bar for i32 {
  type Item<T> = foo;
}
```

Reply:
@nikomatsakis I forgot to add an `impl<T, U> Fn<T, U> for fn<T, U> { }` in that test, I think the (now) corrected test is what you want